### PR TITLE
Generate cache file for SuggestedBindingRedirects

### PR DESCRIPTION
**Pull request by BenVillalobos**
Fixes https://github.com/dotnet/msbuild/issues/5943

### Context
`@(SuggestedBindingRedirects)` is listed as an input for `GenerateBindingRedirects`, but `@(SuggestedBindingRedirects)` _never_ contains a file that exists on disk. This leads to `GenerateBindingRedirects` running for every build.

### Changes Made
`@(SuggestedBindingRedirects)` is now hashed into `obj/$(ProjectFile).SuggestedBindingRedirects.cache` on first build or when there's an update to `SuggestedBindingRedirects`. That file is now used as the input to `GenerateBindingRedirects`, instead of the item that contains no files on disk.

The logic is placed in a new target, `_GenerateSuggestedBindingRedirectsCache` that has `DependsOnTargets="ResolveAssemblyReferences"`

### Testing
Not locally tested yet.

### Notes

### CODE REVIEW

**Comment by BenVillalobos** [comment]
Currently running into an issue where the new suggestedbindingredirects.cache is ALWAYS newer than the app config file. When I skip adding it to FileWrites I no longer see the issue, but I believe it needs to be added to FileWrites for incremental builds.

**Comment by BenVillalobos** [comment]
> Currently running into an issue where the new suggestedbindingredirects.cache is ALWAYS newer than the app config file. When I skip adding it to FileWrites I no longer see the issue, but I believe it needs to be added to FileWrites for incremental builds.

Based on what I'm seeing today, this is only an issue if you don't first clean. When a project is restored or built for the first time, the app config file is generated. 

If you, say, update VS then run a build on a project that _already had_ an app config generated for it, the suggestedbindingredirects cache will be created and compared to the appconfig file. This will then run the target because input is newer than output. The thing is the app config file is only updated when it needs to be. So, if the app config file isn't updated in this scenario, the new cache will _always_ be newer than the appconfig and the target will always run.

TL;DR: Customers might still see the target running at all times until they clean their project. Also tested this on Microsoft.Build.csproj, consecutive builds see the target skipped.

Something I don't fully understand is `GenerateBindingRedirects` showing up twice on projects during consecutive builds.

image

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 1]
```diff
@@ -3617,6 +3618,34 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+
+      <Hash ItemsToHash="@(SuggestedBindingRedirects)">
+        <Output TaskParameter="HashResult" PropertyName="SuggestedBindingRedirectsHash"/>
+      </Hash>
+
+      <WriteLinesToFile Lines="$(SuggestedBindingRedirectsHash)" File="$(SuggestedBindingRedirectsCacheFile)" Overwrite="true" WriteOnlyWhenDifferent="true"/>
+
+      <ItemGroup>
+        <SuggestedBindingRedirectsCache Include="$(SuggestedBindingRedirectsCacheFile)"/>
+        <FileWrites Include="$(SuggestedBindingRedirectsCacheFile)"/>
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:3645]: Do we still want this if we didn't actually overwrite the cache?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -3617,6 +3618,34 @@ Copyright (C) Microsoft Corporation. All rights reserved.
 
   </Target>
 
+  <!--
+    ============================================================
+                                        _GenerateSuggestedBindingRedirectsCache
+    Generate a file used to track suggested binding redirects between builds.
```
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:3624]:
```suggestion
    Generate a file used to track whether suggested binding redirects changed between builds.
```
?

**Review by Forgind — APPROVED** [review]
Looks good to me!

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 3]
```diff
@@ -3617,6 +3618,34 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+    </PropertyGroup>
+
+      <Hash ItemsToHash="@(SuggestedBindingRedirects)">
+        <Output TaskParameter="HashResult" PropertyName="SuggestedBindingRedirectsHash"/>
+      </Hash>
+
+      <WriteLinesToFile Lines="$(SuggestedBindingRedirectsHash)" File="$(SuggestedBindingRedirectsCacheFile)" Overwrite="true" WriteOnlyWhenDifferent="true"/>
+
+      <ItemGroup>
+        <SuggestedBindingRedirectsCache Include="$(SuggestedBindingRedirectsCacheFile)"/>
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:3644]: Why create the item for this instead of just the property?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 1 above)
Comment (rainersigwald) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:3645]: Yes: `@(FileWrites)` is "everything a full build would have written".

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -3617,6 +3618,34 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
+                                        _GenerateSuggestedBindingRedirectsCache
+    Generate a file used to track suggested binding redirects between builds.
+    @(SuggestedBindingRedirects) never contains a file on disk, so create a file
+    that contains a hash of the items to prevent `GenerateBindingRedirects`
+    from running every build.
+
+    See https://github.com/dotnet/msbuild/issues/5943 for details.
+    ============================================================
+    -->
+  <Target Name="_GenerateSuggestedBindingRedirectsCache" Condition="'$(DesignTimeBuild)' != 'true' and '$(BuildingProject)' == 'true'" DependsOnTargets="ResolveAssemblyReferences">
```
Comment (rainersigwald) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:3632]: This feels like a normal bugfix. Do we need a changewave? I think not but want to put it out there.

**Review by rainersigwald — APPROVED** [review]

File: src/Tasks/Microsoft.Common.CurrentVersion.targets (hunk 4 above)
Comment (BenVillalobos) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:3632]: It doesn't warrant a changewave imo. This doesn't break anyone and in the worst case a target is running all of the time, which is the same behavior.
